packages/interaction-tracking/src/InteractionTracking.js

```diff
@@ -0,0 +1,311 @@
+export type Interaction = {|
+  id: number,
+  name: string,
+  timestamp: number,
+|};
+
+export type Subscriber = {
+  // A new interaction has been created via the track() method.
+  onInteractionTracked: (interaction: Interaction) => void,
+
+  // All scheduled async work for an interaction has finished.
+  onInteractionScheduledWorkCompleted: (interaction: Interaction) => void,
+
+  // New async work has been scheduled for a set of interactions.
+  // When this work is later run, onWorkStarted/onWorkStopped will be called.
+  // A batch of async/yieldy work may be scheduled multiple times before completing.
+  // In that case, onWorkScheduled may be called more than once before onWorkStopped.
+  // Work is scheduled by a "thread" which is identified by a unique ID.
+  onWorkScheduled: (interactions: Set<Interaction>, threadID: number) => void,
+
+  // A batch of scheduled work has been cancelled.
+  // Work is done by a "thread" which is identified by a unique ID.
+  onWorkCancelled: (interactions: Set<Interaction>, threadID: number) => void,
+
+  // A batch of work has started for a set of interactions.
+  // When this work is complete, onWorkStopped will be called.
+  // Work is not always completed synchronously; yielding may occur in between.
+  // A batch of async/yieldy work may also be re-started before completing.
+  // In that case, onWorkStarted may be called more than once before onWorkStopped.
+  // Work is done by a "thread" which is identified by a unique ID.
+  onWorkStarted: (interactions: Set<Interaction>, threadID: number) => void,
+
+  // A batch of work has completed for a set of interactions.
+  // Work is done by a "thread" which is identified by a unique ID.
+  onWorkStopped: (interactions: Set<Interaction>, threadID: number) => void,
+};
+
+export type InteractionsRef = {
+  current: Set<Interaction>,
+};
+
+export type Subscribers = Set<Subscriber>;
+export type ScheduledAsyncWorkCounts = Map<Interaction, number>;
+
+const DEFAULT_THREAD_ID = 0;
+
+// Counters used to generate unique IDs.
+let interactionIDCounter: number = 0;
+let threadIDCounter: number = 0;
+
+// Set of currently tracked interactions.
+// Interactions "stack"–
+// Meaning that newly tracked interactions are appended to the previously active set.
+// When an interaction goes out of scope, the previous set (if any) is restored.
+let interactionsRef: InteractionsRef | null = null;
+
+// Listener(s) to notify when interactions begin and end.
+// Note that subscribers are only supported when enableInteractionTrackingObserver is enabled.
+let subscribers: Subscribers | null = null;
+
+// Tracks the number of async operations scheduled for each interaction.
+// Once the number of scheduled operations drops to 0,
+// Interaction subscribers will be notified that the interaction has ended.
+// Note that pending counts are only tracked when enableInteractionTrackingObserver is enabled.
+let scheduledAsyncWorkCounts: ScheduledAsyncWorkCounts | null = null;
```

Comment:
It would be better to store this number on the `Interaction` object itself.

Reply:
Yeah, that could work! Although that would leak it into user space.

Reply:
Underscore! :)

Reply:
Fair enough! 👍